tools/coredump/gosym.go

```diff
@@ -0,0 +1,162 @@
+	test, err = readTestCase(cmd.casePath)
+	if err != nil {
+		return fmt.Errorf("failed to read test case: %w", err)
+	}
+
+	module, addrs, err := goModuleAddrs(test)
+	if err != nil {
+		return fmt.Errorf("failed to find go module addresses: %w", err)
+	}
+
+	goBinary, err := cmd.store.OpenReadAt(module.Ref)
+	if err != nil {
+		return fmt.Errorf("failed to open module: %w", err)
+	}
+	defer goBinary.Close()
+
+	locs, err := goSymbolize(goBinary, addrs)
+	if err != nil {
+		return fmt.Errorf("failed to symbolize: %w", err)
+	}
+
+	for addr, frame := range locs {
+		for _, originFrame := range addrs[addr] {
+			*originFrame = formatSymbolizedFrame(frame, false) + " (" + *originFrame + ")"
+		}
+	}
+
+	return writeTestCaseJSON(os.Stdout, test)
+}
+
+// goModuleAddrs returns the go module and the addresses to symbolize for it
+// mapped to pointers to the frames in c that reference them.
+func goModuleAddrs(c *CoredumpTestCase) (*ModuleInfo, map[libpf.AddressOrLineno][]*string, error) {
+	type moduleAddrs struct {
+		module *ModuleInfo
+		addrs  map[libpf.AddressOrLineno][]*string
+	}
+
+	moduleNames := map[string]*moduleAddrs{}
+	for i, module := range c.Modules {
+		moduleName := filepath.Base(module.LocalPath)
+		if _, ok := moduleNames[moduleName]; ok {
+			return nil, nil, fmt.Errorf("ambiguous module name: %q", moduleName)
+		}
+		moduleNames[moduleName] = &moduleAddrs{
+			module: &c.Modules[i],
+			addrs:  map[libpf.AddressOrLineno][]*string{},
+		}
+	}
+
+	// maxAddrs is the module with the most addresses to symbolize. We use this
+	// as a heuristic to determine which module is the Go module we're
+	// interested in.
+	// TODO(fg) alternatively we could extract all modules and run some check on
+	// them to see if they are go binaries. But this is more complex, so the
+	// current heuristic should be good enough for now.
+	var maxAddrs *moduleAddrs
+	for _, thread := range c.Threads {
+		for i, frame := range thread.Frames {
+			moduleName, addr, err := parseUnsymbolizedFrame(frame)
+			if err != nil {
+				continue
+			}
+
+			moduleAddrs, ok := moduleNames[moduleName]
+			if !ok {
+				return nil, nil, fmt.Errorf("module not found: %q", moduleName)
+			}
+
+			moduleAddrs.addrs[addr] = append(moduleAddrs.addrs[addr], &thread.Frames[i])
+			if maxAddrs == nil || len(moduleAddrs.addrs[addr]) > len(maxAddrs.addrs[addr]) {
+				maxAddrs = moduleAddrs
+			}
+		}
+	}
+	return maxAddrs.module, maxAddrs.addrs, nil
+}
+
+type addrSet[T any] map[libpf.AddressOrLineno]T
+
+func goSymbolize[T any](goBinary io.ReaderAt, addrs addrSet[T]) (map[libpf.AddressOrLineno]*reporter.FrameMetadataArgs, error) {
+	exe, err := elf.NewFile(goBinary)
+	if err != nil {
+		return nil, err
+	}
+
+	lineTableData, err := exe.Section(".gopclntab").Data()
+	if err != nil {
+		return nil, err
+	}
+	lineTable := gosym.NewLineTable(lineTableData, exe.Section(".text").Addr)
+	if err != nil {
+		return nil, err
+	}
+
+	symTableData, err := exe.Section(".gosymtab").Data()
```

Comment:
Another one.

Reply:
https://github.com/open-telemetry/opentelemetry-ebpf-profiler/commit/f9bc8ea6147e7568ed9faec601772025fde1f4a (turns out we don't even need to read this section)

Reply:
Ah yeah: this is only required for very old Go versions. I don't remember which one and can't check my old notes because they live in Elastic repos. 😄

Reply:
The `gosym.NewTable` docs say:

```go
// Starting with Go 1.3, the Go symbol table no longer includes symbol data.
```

So I'm guessing my code doesn't work for Go 1.2 or lower, but I think that's fine for now.